src/server/pkg/work/work.go

```diff
@@ -83,140 +147,203 @@ func (m *Master) Run(ctx context.Context, task *Task) (retErr error) {
-		// Cannot cleanup in this manner, because it messes up subtasks claims when a worker errors.
-		// Cleaning up in the master results with a race condition with workers
-		// waiting on work termination signal and subtask claims to expire.
-		//defer func() {
-		//	// Attempt to cleanup subtask entries.
-		//	if _, err := col.NewSTM(context.Background(), w.etcdClient, func(stm col.STM) error {
-		//		w.subtaskCol.ReadWrite(stm).DeleteAllPrefix(work.Id)
-		//		return nil
-		//	}); err != nil && retErr == nil {
-		//		retErr = err
-		//	}
-		//}()
-		if err := w.processSubtasks(ctx, task); err != nil {
-			return err
-		}
-		// Wait for a deletion event (ttl expired) before attempting to process the subtasks again.
-		return w.subtaskCol.ReadOnly(ctx).WatchOneF(task.Id, func(e *watch.Event) error {
-			return w.processSubtasks(ctx, task)
-		}, watch.WithFilterPut())
-	})
-	if err := eg.Wait(); err != nil {
-		// Work terminated.
-		if ctx.Err() == context.Canceled {
-			return nil
+		case e := <-subtaskWatch.Watch():
+			if e.Type == watch.EventError {
+				return e.Err
+			}
+			var subtaskKey string
+			if err := e.Unmarshal(&subtaskKey, &TaskInfo{}); err != nil {
+				return err
+			}
+			if err := taskEntry.runSubtask(w.subtaskFunc(task, subtaskKey, processFunc)); err != nil {
+				return err
+			}
+		case <-taskEntry.ctx.Done():
+			return taskEntry.ctx.Err()
 		}
-		return err
 	}
-	return nil
 }
 
-func (w *Worker) processSubtasks(ctx context.Context, task *Task) error {
-	for _, subtask := range task.Subtasks {
-		subtaskKey := path.Join(task.Id, subtask.Id)
+func (w *Worker) subtaskFunc(task *Task, subtaskKey string, processFunc ProcessFunc) subtaskFunc {
+	return func(ctx context.Context) error {
+		// (bryce) this should be refactored to have the check and claim in the same stm.
+		// there is a rare race condition that does not affect correctness, but it is less
+		// than ideal because a subtask could get run once more than necessary.
 		subtaskInfo := &TaskInfo{}
-		if err := w.subtaskCol.Claim(ctx, subtaskKey, subtaskInfo, func(ctx context.Context) error {
-			if err := w.processFunc(ctx, task, subtask); err != nil {
+		if _, err := col.NewSTM(ctx, w.etcdClient, func(stm col.STM) error {
+			return w.subtaskCol.ReadWrite(stm).Get(subtaskKey, subtaskInfo)
+		}); err != nil && !col.IsErrNotFound(err) {
+			return err
+		}
+		if subtaskInfo.State == State_RUNNING {
+			if err := w.claimCol.Claim(ctx, subtaskKey, &Claim{}, func(claimCtx context.Context) (retErr error) {
+				defer func() {
+					// If the task was cancelled or the claim was lost, just return with no error.
+					if claimCtx.Err() == context.Canceled {
+						retErr = nil
+						return
+					}
+					subtaskInfo := &TaskInfo{}
+					if _, err := col.NewSTM(claimCtx, w.etcdClient, func(stm col.STM) error {
+						return w.subtaskCol.ReadWrite(stm).Update(subtaskKey, subtaskInfo, func() error {
+							// (bryce) remove when check and claim are in the same stm.
```

Comment:
Something that came up during integration with my branch - this update no longer saves the latest `subtaskInfo.Task.Data` field upon completion of the task, which was behavior I was using to return results to the master.

Reply:
Yeah, that is the behavior we want. Working on a fix now.